activation/nipost_test.go

```diff
@@ -281,73 +310,194 @@ func TestNIPostBuilder_BuildNIPost(t *testing.T) {
 	hash := types.BytesToHash([]byte("anton"))
-	nipost, err := nb.BuildNIPost(context.TODO(), &hash, nil)
+	nipost, _, err := nb.BuildNIPost(context.TODO(), &hash, nil)
 	assert.NoError(err)
 	assert.NotNil(nipost)
 	db := sql.InMemory()
 	assert.Equal(BuilderState{NIPost: &types.NIPost{}}, *nb.state)
 
 	// fail after getting proof ref
-	nb = NewNIPostBuilder(minerID, postProvider, poetProvider, poetDb, db, logtest.New(t))
+	nb = NewNIPostBuilder(minerID, postProvider, []PoetProvingServiceClient{poetProver}, poetDb, db, logtest.New(t))
 	poetDb.errOn = true
-	nipost, err = nb.BuildNIPost(context.TODO(), &hash, nil)
+	nipost, _, err = nb.BuildNIPost(context.TODO(), &hash, nil)
 	assert.Nil(nipost)
 	assert.Error(err)
 
 	// check that proof ref is not called again
-	nb = NewNIPostBuilder(minerID, postProvider, poetProvider, poetDb, db, logtest.New(t))
-	nipost, err = nb.BuildNIPost(context.TODO(), &hash, nil)
+	nb = NewNIPostBuilder(minerID, postProvider, []PoetProvingServiceClient{poetProver}, poetDb, db, logtest.New(t))
+	nipost, _, err = nb.BuildNIPost(context.TODO(), &hash, nil)
 	assert.Nil(nipost)
 	assert.Error(err)
 
 	// fail post exec
-	nb = NewNIPostBuilder(minerID, postProvider, poetProvider, poetDb, db, logtest.New(t))
+	nb = NewNIPostBuilder(minerID, postProvider, []PoetProvingServiceClient{poetProver}, poetDb, db, logtest.New(t))
 	poetDb.errOn = false
 	postProvider.setError = true
 	// check that proof ref is not called again
-	nipost, err = nb.BuildNIPost(context.TODO(), &hash, nil)
+	nipost, _, err = nb.BuildNIPost(context.TODO(), &hash, nil)
 	assert.Nil(nipost)
 	assert.Error(err)
 
 	// fail post exec
-	nb = NewNIPostBuilder(minerID, postProvider, poetProvider, poetDb, db, logtest.New(t))
+	nb = NewNIPostBuilder(minerID, postProvider, []PoetProvingServiceClient{poetProver}, poetDb, db, logtest.New(t))
 	poetDb.errOn = false
 	postProvider.setError = false
 	// check that proof ref is not called again
-	nipost, err = nb.BuildNIPost(context.TODO(), &hash, nil)
+	nipost, _, err = nb.BuildNIPost(context.TODO(), &hash, nil)
 	assert.NotNil(nipost)
 	assert.NoError(err)
 
 	assert.Equal(3, postProvider.called)
 	// test state not loading if other challenge provided
-	poetProvider.EXPECT().PoetServiceID(gomock.Any()).Return([]byte{}, nil)
-	poetProvider.EXPECT().Submit(gomock.Any(), gomock.Any()).Return(&types.PoetRound{}, nil)
+	poetProver.EXPECT().PoetServiceID(gomock.Any()).Return([]byte{}, nil)
+	poetProver.EXPECT().Submit(gomock.Any(), gomock.Any()).Return(&types.PoetRound{}, nil)
 	hash2 := types.BytesToHash([]byte("anton1"))
-	nipost, err = nb.BuildNIPost(context.TODO(), &hash2, nil)
+	nipost, _, err = nb.BuildNIPost(context.TODO(), &hash2, nil)
 	assert.Equal(4, postProvider.called)
 
 	assert.NotNil(nipost)
 	assert.NoError(err)
 }
 
+func createMockPoetService(t *testing.T, id []byte) PoetProvingServiceClient {
+	t.Helper()
+	poet := mocks.NewMockPoetProvingServiceClient(gomock.NewController(t))
```

Comment:
By using this helper method multiple `gomock` Controllers are instantiated within the same test.

In my opinion this helper can be inlined. It only saves 3 lines of code per call and is called 4 times, but hides test details outside of the test.

Reply:
> By using this helper method multiple gomock Controllers are instantiated within the same test.
Yes, but it doesn't hurt.

IMO, it makes the test code more readable 🤷 .